### ISSUE TITLE
Implement constraint solving

### ISSUE THREAD

**Issue by jlapacik** [influxdata/flux#1958]
Given a set of constraints, solve those constraints, producing a substitution or a unification error. Note this is needed in order to generate constraints for a semantic block.

**Comment by jlapacik** [comment]
To be more concrete, this issue is about implementing a function similar to the one below:
```Rust
fn solve(cons: Vec<Constraint>) -> Result<Substitution, Error> {
    ...
}
```
The algorithm is as follows:
1. Pop constraint off stack
2. Apply current substitution to both sides of constraint.
3. Unify both sides of constraint producing another substitution.
4. Merge this substitution into the original one.
5. Repeat until constraint stack is empty.

Note the return type of `solve` may be slightly different than the one above to account for Kind constraints.

**Comment by jlapacik** [comment]
Closed by #2061.
